tiny_dnn/parameter.h

```diff
@@ -0,0 +1,125 @@
+/*
+    Copyright (c) 2013, Taiga Nomi and the respective contributors
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/tensor.h"
+#include "tiny_dnn/util/util.h"
+
+namespace tiny_dnn {
+
+enum class parameter_type : int8_t { weight = 0x0001, bias = 0x0002 };
+
+class parameter {
+ public:
+  /**
+   * Initializes an empty parameter taking in the dimensions of weights and
+   * biases of a layer. Currently supported for maximum 4-dimensions, and
+   * stored as a flat ``Tensor``.
+   *
+   * todo (karandesai) : generalize to n-dimensions
+   * todo (karandesai) : add an n-dimensional view for easy indexing
+   *
+   * @param width      [in] filter width
+   * @param height     [in] filter height
+   * @param depth      [in] filter depth / input channels
+   * @param n_fmaps    [in] number of output feature maps
+   * @param type       [in] whether parameter is a weight or a bias
+   * @param trainable  [in] whether parameter will be updated while training or
+   * not
+   */
+  parameter(serial_size_t width,
+            serial_size_t height,
+            serial_size_t depth,
+            serial_size_t n_fmaps,
+            parameter_type type,
+            bool trainable = true)
+    : type_(type),
+      shape_(width, height, depth),
+      n_fmaps_(n_fmaps),
+      trainable_(trainable),
+      data_({size()}),
+      grad_({1, size()}) {}
+
+  shape3d shape() { return shape_; }
+
+  size_t size() { return shape_.size() * n_fmaps_; }
```

Comment:
I guess that `Tensor` has a routine for that, if that's not the case implement it